src/html/XHPHelpers.php

```diff
@@ -0,0 +1,183 @@
+          'support the "id" attribute. The best way to do this is to inherit '.
+          'the HTML attributes from the element your component will render '.
+          'into.',
+        );
+      }
+    }
+    return (string)$id;
+  }
+
+  /*
+   * Fetches the "id" attribute, will generate a unique value if not set.
+   */
+  final public function getID(): string {
+    return $this->requireUniqueID();
+  }
+
+  /*
+   * Copies all attributes that are set on $this and valid on $target to
+   * $target.
+   */
+  final public function copyAllAttributes(
+    :xhp $target,
+  ): void {
+    $this->transferAttributesImpl($target, Set{});
+  }
+
+  /*
+   * Copies only the non-HTML attributes that are set on $this and valid on
+   * $target to $target.
+   */
+  final public function copyCustomAttributes(
+    :xhp $target,
+  ): void {
+    $this->transferAttributesImpl($target);
+  }
+
+  /*
+   * Copies all attributes except those specified that are set on $this and
+   * valid on $target to $target.
+   */
+  final public function copyAttributesExcept(
+    :xhp $target,
+    Set<string> $ignore,
+  ): void {
+    $this->transferAttributesImpl($target, $ignore);
+  }
+
+  /*
+   * Transfers all attributes that are set on $this and valid on $target to
+   * $target. This will unset all transfered attributes from $this.
+   */
+  final public function transferAllAttributes(
+    :xhp $target,
+  ): void {
+    $this->transferAttributesImpl($target, Set{}, true);
+  }
+
+  /*
+   * Transfers only the non-HTML attributes that are set on $this and valid on
+   * $target to $target. This will unset all transfered attributes from $this.
+   */
+  final public function transferCustomAttributes(
+    :xhp $target,
+  ): void {
+    $this->transferAttributesImpl($target, null, true);
+  }
+
+  /*
+   * Transfers all attributes except those specified that are set on $this and
+   * valid on $target to $target. This will unset all transfered attributes from
+   * $this.
+   */
+  final public function transferAttributesExcept(
+    :xhp $target,
+    Set<string> $ignore,
+  ): void {
+    $this->transferAttributesImpl($target, $ignore, true);
+  }
+
+  /*
+   * The grunt work of transfering attributes. Don't call this function
+   * directly. Instead, use one of the transfer/copy flavors above.
+   */
+  final private function transferAttributesImpl(
+    :xhp $target,
+    ?Set<string> $ignore = null,
+    bool $remove = false,
+  ): void {
+    if ($ignore === null) {
+      $ignore = :xhp:html-element::__xhpAttributeDeclaration();
+    } else {
+      $ignore = array_fill_keys($ignore->toArray(), true);
+    }
+
+    $compatible = new Map($target::__xhpAttributeDeclaration());
+    $transferAttributes = array_diff_key($this->getAttributes(), $ignore);
+    foreach ($transferAttributes as $attribute => $value) {
+      if ($compatible->containsKey($attribute))
+          || ReflectionXHPAttribute::IsSpecial($attribute)) {
```

Comment:
This is what Travis is complaining about - there's too many closing parens on the previous line